src/restricted-endpoints/replication/changes/changes.controller.ts

```diff
@@ -31,31 +33,68 @@ export class ChangesController {
    * @param user
    */
   @Get(':db/_changes')
-  changes(
+  async changes(
     @Param('db') db: string,
-    @Query() params,
     @User() user: UserInfo,
-  ): Observable<ChangesResponse> {
-    return this.couchdbService
-      .get<ChangesResponse>(db, '_changes', {
-        ...params,
-        include_docs: true,
-      })
-      .pipe(map((res) => this.filterChanges(res, user)));
+    @Query() params?: ChangesParams,
+  ): Promise<ChangesResponse> {
+    const ability = this.permissionService.getAbilityFor(user);
+    const change = { results: [] } as ChangesResponse;
+    let since = params?.since;
+    while (true) {
+      const res = await this.getPermittedChanges(
+        db,
+        { ...params, since },
+        ability,
+      );
+      // missing changes till limit
+      const missing = (params?.limit ?? Infinity) - change.results.length;
+      // overflow changes of this request
+      const discarded = Math.max(res.results.length - missing, 0);
+      change.results.push(...res.results.slice(0, missing));
+      if (change.results.length > 0) {
+        change.last_seq = change.results[change.results.length - 1].seq;
+      }
+      change.pending = res.pending + discarded;
+      if (
+        !params?.limit ||
+        change.pending === 0 ||
+        change.results.length >= params.limit
+      ) {
+        // enough changes found or none left
+        break;
+      }
+      since = res.last_seq;
+    }
+    if (params?.include_docs !== 'true') {
+      // remove doc content if not requested
+      change.results = change.results.map((c) => omit(c, 'doc'));
+    }
+    return change;
+  }
+
+  getPermittedChanges(
+    db: string,
+    params: ChangesParams,
+    ability: DocumentAbility,
+  ): Promise<ChangesResponse> {
+    return firstValueFrom(
+      this.couchdbService
+        .get<ChangesResponse>(db, '_changes', {
+          ...params,
+          include_docs: true,
+        })
+        .pipe(map((res) => this.filterChanges(res, ability))),
+    );
   }
 
   private filterChanges(
     changes: ChangesResponse,
-    user: UserInfo,
+    ability: DocumentAbility,
   ): ChangesResponse {
-    const ability = this.permissionService.getAbilityFor(user);
-    changes.results = changes.results
-      .filter((change) => this.canReadDoc(change.doc, ability))
-      .map((change) => omit(change, 'doc'));
+    changes.results = changes.results.filter(
+      ({ doc }) => doc._deleted || ability.can('read', doc),
+    );
```

Comment:
```suggestion
    changes.results = changes.results.filter(
      ({ doc }) => doc._deleted || ability.can('read', doc),
    ).map((c) => ability.can('read', c.doc) ? c : omit(c, 'doc'));
```

Reply:
Good idea, I adjusted it a little bit to not have a inconsistent structure at the end (some changes have a `doc` property while others don't)